### DISCUSSION TITLE
Type-dependent doctor checks

### DISCUSSION THREAD

**Discussion by TheTaoOfSu** [papis/papis#691] [Ideas]
This is mainly inspired by the keys-exist check. It allows you to define in your settings a list of keys that entries must possess so that any missing keys are brought to your attention when the check is run. This is useful for verifying some universal basics, but there are keys that I want to ensure are there that depend on the entry's type. For example, I require that all entries have the title, author, ref, date (of publication), and type, but if it's an online reference, it should include url and urldate for proper citing, if its type is book, it should include some basic information about the publisher at a minimum, and if it's a journal article, it should include the name of the journal it's published in.

I may be able to aid in implementing this functionality, but my programming experience is limited, so I feel ill-equipped to make decisions on things like how users would set it in settings, any extra command line flags this may necessitate, etc. With a basic framework laid out for me, simply adding the functionality doesn't seem too hard.

**Comment by alexfikl** [comment]
In the development version, there's a new check called `biblatex-required-keys` that checks something like that for standard BibTex types as described in the [BibLaTeX manual](https://mirrors.nxthost.com/ctan/macros/latex/contrib/biblatex/doc/biblatex.pdf). See here for an updated list:
https://papis.readthedocs.io/en/latest/commands/doctor.html
https://papis.readthedocs.io/en/latest/developer_reference.html#papis.commands.doctor.biblatex_required_keys_check

Would that work for you or do you need some more customization?

**Reply by TheTaoOfSu** [reply]
That... Actually looks perfect, so you guys are clearly a step ahead of me. I just want to make sure that all entries of a certain type have some subset of keys at a minimum, where the subset is the same for all entries of a given type but may be different for all entries of a _different_ type, e.g., all books need publisher, all online need url and urldate, all journal articles need the journal name, etc., but none care what the other need, so books don't care if they have url, urldate, or a journal name, etc.

**Reply by alexfikl** [reply]
Yeah, that's how that check works for the standard BibTeX types, e.g.:
* `article` requires `author`, `title`, `journal/journaltitle`, and `year / date`
* `book` requires `author`, `title`, and `year/date`.
* `online` requires `author/editor`, `title`, `year/date`, and `doi/eprint/url`. They seem to prefer `date` to `urldate`, but we don't need to be as strict.

and similarly for the other recognized types. You can check the [manual](https://mirrors.nxthost.com/ctan/macros/latex/contrib/biblatex/doc/biblatex.pdf) in `Section 2.1 Entry Types` for a complete list. We could make that a bit more configurable, if needed, since some other BibTeX fields can also be extended with user types and fields.
